common/persistence/cassandra/metadata_store.go

```diff
@@ -161,23 +168,37 @@ func (m *MetadataStore) CreateNamespaceInV2Table(
 
 	previous := make(map[string]interface{})
 	applied, iter, err := m.session.MapExecuteBatchCAS(batch, previous)
+	deleteOrphanNamespace := func() {
+		// Delete namespace from `namespaces_by_id`
+		if errDelete := m.session.Query(templateDeleteNamespaceQuery, request.ID).WithContext(ctx).Exec(); errDelete != nil {
+			m.logger.Warn("Unable to delete orphan namespace record. Error", tag.Error(errDelete))
+		}
+	}
+
 	if err != nil {
 		return nil, serviceerror.NewUnavailable(fmt.Sprintf("CreateNamespace operation failed. Inserting into namespaces table. Error: %v", err))
 	}
+
 	defer func() { _ = iter.Close() }()
 
 	if !applied {
-		// Namespace already exist.  Delete orphan namespace record before returning back to user
-		if errDelete := m.session.Query(templateDeleteNamespaceQuery, request.ID).WithContext(ctx).Exec(); errDelete != nil {
-			m.logger.Warn("Unable to delete orphan namespace record. Error", tag.Error(errDelete))
-		}
 
-		if id, ok := previous["Id"].([]byte); ok {
-			msg := fmt.Sprintf("Namespace already exists.  NamespaceId: %v", primitives.UUIDString(id))
+		// if conditional failure is due to a duplicate name in namespaces table
+		if name, ok := previous["name"]; ok && name == request.Name {
+			if id, ok := previous["id"]; ok && gocql.UUIDToString(id) != request.ID {
+				// Delete orphan namespace record before returning back to user
+				deleteOrphanNamespace()
+			}
+			msg := fmt.Sprintf("Namespace already exists.  NamespaceId: %v", id)
 			return nil, serviceerror.NewNamespaceAlreadyExists(msg)
+
 		}
 
-		return nil, serviceerror.NewNamespaceAlreadyExists("CreateNamespace operation failed because of conditional failure.")
+		// If namespace does not exist already and applied is false,
+		// notification_version does not match our expectations and it's conditional failure.
+		// Delete orphan namespace record before returning back to user
+		deleteOrphanNamespace()
```

Comment:
this is not required since first table can handle retry now?

Reply:
- I am not against doing this, mainly want to confirm my understanding. Please feel free to land PR.
- From guarantee perspective, the deletion can also fail, so there's always a possibility there will be an orphan entry in the first table.

Reply:
I agree, both approaches are fine I guess. I am a fan of cleaning up and leave things as they were in case of error, so let me keep it as is then.